sender_test.go

```diff
@@ -0,0 +1,31 @@
+package amqp
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestClosedSenderReturnsErrClosed(t *testing.T) {
+	// this feels a bit _too_ fake, should revisit.
+	link, err := newLink(newSession(nil, 0), &Receiver{}, nil)
+	require.NoError(t, err)
+
+	sender := &Sender{link: link}
+
+	// simulate a detach happening before the send()
+	// link.muxDetach()
+	close(link.done)
```

Comment:
Can we not call `sender.Close()` instead?

Reply:
This brings up an interesting question that I didn't realize until you asked this.

TLDR: I think this code is 'correct', but only because of the way the code is _currently_.

Longer:

There are (apparently) two channels that indicate a closed state with the `link`. `done` and `close`. 

* `close(l.done)` gets called if the service requests you close (ie: detach) or if the server sends back an error, in which case we do the detach. I renamed it to `detached` just now because that's what it's meant to represent. 
* `close(l.close)` gets called if a caller calls l.Close() explicitly. So it represents a user-initiated close.

I think there's value in keeping these separate (a user calling Close() means the link should not reopen, whereas the service asking you to detach does NOT (ie, detach should be considered recoverable)).

What we could do is actually create an ErrLinkDetached or similar, that lets you distinguish these programatically. That way our recovery code can still take advantage of this, but also do the right thing if the link was closed (ie - don't reopen it!).

Reply:
Just FYI: I added a function (link.Check) that returns either ErrLinkClosed or ErrLinkDetached when called and migrated my code into there. I think it's doing the right thing in a better way (check that the link is valid) and we can expand it as we find more cases (although I don't anticipate any).